LeaderboardBackend/Authorization/UserTypeAuthorizationHandler.cs

```diff
@@ -0,0 +1,106 @@
+using LeaderboardBackend.Models.Entities;
+using LeaderboardBackend.Services;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.IdentityModel.Tokens;
+using System.IdentityModel.Tokens.Jwt;
+
+namespace LeaderboardBackend.Authorization;
+
+public class UserTypeAuthorizationHandler : AuthorizationHandler<UserTypeRequirement>
+{
+	private readonly IConfiguration _config;
+	private readonly JwtSecurityTokenHandler _jwtHandler;
+	private readonly TokenValidationParameters _jwtValidationParams;
+	private readonly IUserService _userService;
+
+	public UserTypeAuthorizationHandler(
+		IConfiguration config,
+		IModshipService modshipService,
+		IUserService userService
+	)
+	{
+		_config = config;
+		_jwtHandler = JwtSecurityTokenHandlerSingleton.Instance;
+		_jwtValidationParams = TokenValidationParametersSingleton.Instance(config);
+		_userService = userService;
+	}
+
+	protected override Task HandleRequirementAsync(
+		AuthorizationHandlerContext context,
+		UserTypeRequirement requirement
+	)
+	{
+		if (!CanGetJwt(context, out string token) || !CanValidateJwt(token))
+		{
+			return Task.CompletedTask;
+		}
+
+		User? user = _userService.GetUserFromClaims(context.User).Result;
+
+		if (user is null || !Handle(user, context, requirement))
+		{
+			// FIXME: Work out how to fail as a ForbiddenResult.
+			context.Fail();
+			return Task.CompletedTask;
+		}
+
+		context.Succeed(requirement);
+
+		return Task.CompletedTask;
+	}
+
+	private bool Handle(
+		User user,
+		AuthorizationHandlerContext context,
+		UserTypeRequirement requirement
+	) => requirement.Type switch
+	{
+		UserTypes.Admin => IsAdmin(user),
+		UserTypes.Mod => IsMod(user),
+		UserTypes.User => true,
+		_ => false,
+	};
+
+	private bool IsAdmin(User user) => user.Admin;
+
+	// FIXME: Users don't get automagically populated with Modships when on creation of the latter.
+	private bool IsMod(User user) => user.Modships?.Count() > 0;
+
+	private bool CanGetJwt(AuthorizationHandlerContext context, out string token)
```

Comment:
Naming convention in C# for this pattern is usually `TryGetJwt`. I'd probably recommend taking the name another step too, `TryGetJwtFromHttpContext`.